src/api/lexer/lexing.rs

```diff
@@ -0,0 +1,83 @@
+// This file is part of rust-web/twig
+//
+// For the copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+use api::Position;
+use api::tokens::TokenRef;
+
+use api::lexer::{ Lexer, LexingResult };
+use api::error::{ SyntaxErrorCode, UnexpectedToken };
+
+/// TokenRef wrapper for `Lexer` that additionaly has position.
+#[derive(Debug)]
+pub struct ItemRef<'t> {
+    pub token: TokenRef<'t>,
+    pub position: Position,
+}
+
+/// Lexer token iterator.
+///
+/// 'i is iteration lifetime, or "one use of lexer".
+/// 't is template lifetime. It will live longer than this iteration.
+pub struct Tokens<'i, 't> {
+    /// Position of the next token to get.
+    next_pos: Position, // temporary field until I get cursor in.
+    env: &'i Lexer,
+    code: &'t str,
+}
+
+impl<'i, 't> Iterator for Tokens<'i, 't> {
+    type Item = LexingResult<ItemRef<'t>>;
+
+    fn next(&mut self) -> Option<LexingResult<ItemRef<'t>>> {
```

Comment:
TL;DR at the end ^^

Too bad rust does not have guaranteed tail call elimination. I did some investigations, what we could do instead.

First, I analyzed the nesting of sub-calls in my code and twigphp to look for any possibly infinite recursions.
The short answer is, there are definitely two possible recursion loops:
* state::Data::tokenize (aka lex_data) calls itself (luckily in a tail position)
* lex_expression calls itself (a) directly and possibly via (b) lex_expression -> lex_string -> lex_interpolation -> lex_expression

These recursions are separately closed in a sense they can be rewritten individually like

```rust
'lex_data: loop {
    // do something
    if(/**/) { continue 'lex_data /* mimic tail-call */ }
    // ..
    if(/**/) { break 'lex_data /* or return */ }
    // ..
}
```

However there might be another recursion related to interpolation in double-quoted strings. This boils down to "can an expression/variable contain a string that contains a variable that contains another string that .."

Here are some details. I denote subcalls by "->", where tail-/non-tail-positions are marked with "*" and "_". No sub-calls: "()".

```
Initial -> *Data
Data -> *Data + *Final + _lex_comment + _lex_verbatim_data + _Block + _Expression
Final -> ()
lex_comment -> ()
lex_verbatim_data -> ()
Block -> *lex_expression
Expression/Variable -> *lex_expression
lex_expression -> *String + *PARENT (e.g. Block, Expression/Variable or Interpolation -> continue 'PARENT )
String -> *Interpolation
Interpolation -> *lex_expression (-> we can only reach this inside another lex_expression, so we could just `continue 'lex_expression` in our loop version - if it wasn't for the call from lex_expression to *PARENT -> this loop might be the most difficult one)
```

The good news is: sub-calls in non-tail-positions "_" don't create recursions (they only occur in Data, so this is easy to analyse).

* The lex_data-recursion is easy to rewrite (see above).
* And I am very confident the lex_expression recursion can be rewritten as a flat loop, too - however we probably want to share the code between Block and Expression/Variable.

We can't break a loop of the caller from inside a function - even a generic function is not flexible enough - but macros can. For readability I put the macros to the end - but for compilation they must be defined first.
```rust
'lex_block: loop {
  // ..
  lex_expression!("block successfully parsed", break 'lex_block, continue 'lex_block); // not in tail position - but no guarantee needed, because not recursive
  continue 'lex_data // equivalent to a tail-call guarantee
  // ..
}

// shared code via macro - a bit nasty but it works(tm)
macro_rules! lex_expression {
    ( $message:expr, $break_parent:stmt, $continue_parent:stmt ) => {
        'lex_expression: loop {
		// do something
		println!($message);
		$break_parent;
		unreachable!(); // just to check the break_parent stmt really jumps somewhere else
		// ...
                // we could embedd the lex_string logic completely in this macro, or delegate to a sub-macro
		lex_string!("some argument", break 'lex_expression, continue 'lex_expression)
                //
        }
    };
}

// shared code via macro - a bit nasty but it works(tm)
macro_rules! lex_string {
    ( $message:expr, $break_parent:stmt, $continue_parent:stmt ) => {
        'lex_string: loop {
		// do something
		println!($message);
		$continue_parent;
		unreachable!(); // just to check the break_statement really jumps somewhere else
		// ...
        }
    };
}
```

I know this macro-thing probably feels a bit weird. But if we want to factor away most function calls in lexer, it seems possible. The only thing that needs a closer look is the possible expression-interpolation-recursion (this is where the stack could still blow up, if we can't reduce it to a flat loop). But at this point even I begin to question if it is really worth it... 

# TL;DR

So finally I arrive at the point to admit "let's just proceed with some enum-based match-branching in a loop - which would include the iterator pattern you are suggesting". ;-)